### PR TITLE
[EVM] Implement large constants unfolding after stackification

### PR DESCRIPTION
**Pull request by PavelKopyl**
A large constant is replaced with a sequence of
operations to reduce a bytecode size.

For example,

  0xFFFFFFFF00000000000000000000000000000000000000000000000000000000

can be replaced with

  0xFFFFFFFF << 224
  
This pass runs after stackification to prevent additional stack-too-deep errors, which could otherwise arise from increased register pressure caused by a higher number of live values.

That said, there may be potential benefits to performing unfolding before stackification, since identical immediate values (if present) could be deduplicated during stackification. It would be reasonable to revisit this approach after resolving the stack-too-deep issues introduced by LiteralSlots.

Further possible enhancements concerning large constants (not limited to this patch):
 - Disable unfolding inside a loop when optimizing for speed
 - Spill frequently used constants to memory. This can be done at the module level, allowing all functions to access a shared "constant pool" in memory.
# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+  unsigned ValLen = Val.getActiveBits();
+  APInt AbsVal = Val.abs();
+  unsigned AbsValLen = AbsVal.getActiveBits();
+  bool IsMask = ((Ones + LeadZ + TrailZ) == Imm.getBitWidth());
+  assert(ValLen == (Imm.getBitWidth() - TrailZ - LeadZ));
+  assert(Val.isNegative());
+
+  // The upper limit for both size and instruction count.
+  constexpr unsigned MaxCost = /* PUSH32 size*/ 33 * 2;
+  SmallVector<std::pair<std::function<std::pair<unsigned, unsigned>(void)>,
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Is it possible to name these structures (outer and inner) for better readability?

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <functional>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "evm-constant-unfolding"
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp:42]: Do we want to add statistics for this patch?

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+        // Expressing total cost in terms of both size and number of operations
+        // enables more performance-friendly transformation choices. For
+        // example, consider the following two options:
+        //
+        //   [size: 7, OpNum: 5]
+        // vs
+        //   [size: 8, OpNum: 3]
+        //
+        // The second option is likely preferable, as the operation count is
+        // notably lower while the size remains nearly the same.
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Do we have any perf numbers for this?

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+      dbgs() << "Applying optimal transform : [size: " << Cost.first
+             << ", OpNum: " << Cost.second << "]\n";
+    });
+  } else {
+    unsigned OrigSize = (alignTo(Imm.getActiveBits(), 8) / 8) + 1;
+    // When optimizing for speed, only unfold constants if it reduces the size
+    // by a factor of at least 3 and keeps the instruction count to 3 or fewer.
+    // This seems to be an effective heuristic that avoids significantly
+    // impacting performance.
+    if (OrigSize / Cost.first >= 3 && Cost.second <= 3) {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: How this 'magic number' was calculated? Could we extract it to a named variable for better readability?

File: llvm/test/CodeGen/EVM/constant-unfolding.ll [hunk 5]
```diff
@@ -0,0 +1,267 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Comment (akiramenai) [inline llvm/test/CodeGen/EVM/constant-unfolding.ll:1]: Could we add a test that we do not try to unfold small constants?

File: llvm/test/CodeGen/EVM/constant-unfolding-Oz.ll [hunk 6]
```diff
@@ -0,0 +1,304 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -O3 < %s | FileCheck %s
+target datalayout = "E-p:256:256-i256:256:256-S256-a:256:256"
+target triple = "evm-unknown-unknown"
+
+declare void @llvm.evm.return(ptr addrspace(1), i256) noreturn
+
+; 0x4e487b7100000000000000000000000000000000000000000000000000000000
```
Comment (akiramenai) [inline llvm/test/CodeGen/EVM/constant-unfolding-Oz.ll:10]: I'd add a short clarification why we expect here (and below) one or another transformation done.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+  }
+
+  if (Changed)
+    MI.eraseFromParent();
+
+  return Changed;
+}
+
+bool static isPUSH(const MachineInstr &MI) {
+  switch (MI.getOpcode()) {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Can we tablegen it?

**Comment by akiramenai** [comment]
@kpv — sorry for the slow review. I’m looking at the patch now and should finish tonight or tomorrow.

I suggest two improvements:

1. **TableGen the `isPUSH` switch.** You can also TableGen aux functions from `EVMTargetInfo.cpp`.  
2. **Handle minsize functions and cold code.** For example:
   ```cpp
   auto &MBFI = getAnalysis<MachineBlockFrequencyInfo>();
   BlockFrequency Freq = MBFI.getBlockFreq(&MBB);
   if (Freq < SomeThreshold) {
     // treat MBB as cold
   }

**Comment by akiramenai** [comment]
Also, do you know why do we have gas improvements in this patch? It's not what I would expect if we replace one instruction with multiple.

**Comment by PavelKopyl** [comment]
> Also, do you know why do we have gas improvements in this patch? It's not what I would expect if we replace one instruction with multiple.

As I understand that's because we reduce the deployment bytecode size, for which we are charged less in gas at deploy time. It seems that, for CI tests, the savings from reduced deployment costs outweigh the increased gas usage at runtime.

**Comment by akiramenai** [comment]
So we count deployment costs towards `Gas (-%)` category, don't we? But what is the assumption on how many time the contract will be run? Coupling deploy-time and run-time gas looks confusing to me.
@antonbaliasnikov are we correct? Is there a way we can decouple?
@PavelKopyl for this particular patch it would be great to understand what happened to runtime gas. Is there a way to get this data? Maybe with the performance stand if the CI can't help now.

**Comment by akiramenai** [comment]
Also, do `cycles` reflect reality in some shape or form? I remember there were some concerns in how we measure cycles via the interpreter.
They went down as well.

**Comment by PavelKopyl** [comment]
> So we count deployment costs towards `Gas (-%)` category, don't we? But what is the assumption on how many time the contract will be run? Coupling deploy-time and run-time gas looks confusing to me. @antonbaliasnikov are we correct? Is there a way we can decouple? @PavelKopyl for this particular patch it would be great to understand what happened to runtime gas. Is there a way to get this data? Maybe with the performance stand if the CI can't help now.

>So we count deployment costs towards `Gas (-%)` category, don't we? But what is the assumption on how many time the contract will be run? Coupling deploy-time and run-time gas looks confusing to me.

Yes, these are good questions. I hope @antonbaliasnikov will clarify how exactly that values are calculated. I'll also will check `benchmark-analyzer` tool.

 >it would be great to understand what happened to runtime gas. Is there a way to get this data? Maybe with the performance stand if the CI can't help now.

I think we can tweak `benchmark-analyzer` to just ignore lines containing #deployer to take into account only the runtime values.

**Comment by PavelKopyl** [comment]
> Also, do `cycles` reflect reality in some shape or form? I remember there were some concerns in how we measure cycles via the interpreter. They went down as well.

Right, as @hedgar2017 explained to me, we should ignore `cycles`, as they do not corresponds to actually EVM cycles.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+  unsigned AbsValLen = AbsVal.getActiveBits();
+  bool IsMask = ((Ones + LeadZ + TrailZ) == Imm.getBitWidth());
+  assert(ValLen == (Imm.getBitWidth() - TrailZ - LeadZ));
+  assert(Val.isNegative());
+
+  // The upper limit for both size and instruction count.
+  constexpr unsigned MaxCost = /* PUSH32 size*/ 33 * 2;
+  SmallVector<std::pair<std::function<std::pair<unsigned, unsigned>(void)>,
+                        std::function<bool(void)>>>
+      Transformations = {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: I would prefer extracting transformation as functions, name them properly and attach documentation. If lamas are needed to capture, we can capture and make a call.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    Changed = OptTransform->second();
+    LLVM_DEBUG({
+      dbgs() << "Applying optimal transform : [size: " << Cost.first
+             << ", OpNum: " << Cost.second << "]\n";
+    });
+  } else {
+    unsigned OrigSize = (alignTo(Imm.getActiveBits(), 8) / 8) + 1;
+    // When optimizing for speed, only unfold constants if it reduces the size
+    // by a factor of at least 3 and keeps the instruction count to 3 or fewer.
+    // This seems to be an effective heuristic that avoids significantly
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: > This seems to be an effective heuristic

Could you give some objective data in support?

**Review by akiramenai — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Done.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 2 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp:42]: Done.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 3 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: I've just checked on small tests that the comparison in non-OptForSize mode:
`(CostA.first + CostA.second) < (CostB.first + CostB.second)`
is more adequate than the comparison for OptForSize mode.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 4 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: I got it experimentally on our CI tests, I've converted these two magic numbers into options.

File: llvm/test/CodeGen/EVM/constant-unfolding.ll (hunk 5 above)
Comment (PavelKopyl) [reply llvm/test/CodeGen/EVM/constant-unfolding.ll:1]: Done.

File: llvm/test/CodeGen/EVM/constant-unfolding-Oz.ll (hunk 6 above)
Comment (PavelKopyl) [reply llvm/test/CodeGen/EVM/constant-unfolding-Oz.ll:10]: We choose a concrete transformation based on the cost calculations and choosing the cheapest one. Do you mean we need to specify costs for each possible transformation for each test?

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 7 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Yeap, done.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 8 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Sure, done.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 9 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Please, find below the aave-v3 gas/size changes after enabling this optimization.

[aave-const-unfolding.txt](https://github.com/user-attachments/files/20909219/aave-const-unfolding.txt)

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 3 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: I could check different metrics, but this should be done on real-life projects.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 9 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Please, also note that, as I understand, it's not possible to completely split size/gas costs, because when we execute CREATE opcode, the consumed amount of gas (runtime) depends on a size of a new contract.

**Review by akiramenai — APPROVED** [review]
LGTM, thank you

**Comment by hedgar2017** [comment]
@akiramenai @PavelKopyl needs squashing?

**Comment by PavelKopyl** [comment]
> @akiramenai @PavelKopyl needs squashing?

Done.
